### PR TITLE
Use floating point exceptions to catch more errors in the testsuite.

### PR DESCRIPTION
**Pull request by bangerth**
This is inspired by @tjhei 's use here: https://github.com/geodynamics/aspect/pull/441

There are a number of tests that currently fail:
```
        11 - algorithms/newton_01.debug (Failed)
        821 - base/timer_02.debug (Failed)
        841 - base/log_nan.debug (Failed)
        881 - base/is_finite.debug (Failed)
        1000 - base/function_parser_07.debug (Failed)
        1135 - base/is_finite_complex.debug (Failed)
        1266 - base/function_parser_06.debug (Failed)
        2916 - codim_one/integrate_log_2.debug (Failed)
        2932 - codim_one/transform_01.debug (Failed)
        3022 - codim_one/bem_integration.debug (Failed)
```
See also #963 and #964. The `is_finite` tests will need to be treated separately. The `codim_*` tests are real problems. So is the timer test.

### CODE REVIEW

**Comment by tamiko** [comment]
Good idea, but ```feenableexcept``` is glibc exclusive - and even there not present on all platforms. Thus, we need a proper configure check in ```cmake/check/check_02_system_features.cmake```.

**Comment by tjhei** [comment]
See https://github.com/geodynamics/aspect/pull/473 for the configure check in ASPECT (and also using the c version instead of the cxx11 version)

**Comment by bangerth** [comment]
I think the current set of failing tests with this patch is
```
       825 - base/timer_02.debug (Failed)
        845 - base/log_nan.debug (Failed)
        885 - base/is_finite.debug (Failed)
        2920 - codim_one/integrate_log_2.debug (Failed)
        2936 - codim_one/transform_01.debug (Failed)
        3026 - codim_one/bem_integration.debug (Failed)
```

**Comment by bangerth** [comment]
And actually also `lac/linear_operator_03.debug`.

**Comment by bangerth** [comment]
This will also need the patches from https://github.com/geodynamics/aspect/pull/530 , see https://github.com/geodynamics/aspect/pull/500 for the reason.

**Comment by tamiko** [comment]
What is the current status?

**Comment by bangerth** [comment]
No news. I should on occasion try it again, I suppose.

**Comment by bangerth** [comment]
There are now open PRs or issues (#1058) for all current test failures. Let's revisit the issue when they are all addressed.

**Comment by bangerth** [comment]
Current fails:
```
base/timer_02.debug
base/log_nan.debug
base/is_finite.debug
base/is_finite_complex.debug
```
The problem with the latter 3 is that (helpfully :-( ) the function `isnan` traps with a floating point exception when you call it wit a signalling NaN. Example:
```
#include <fenv.h>
#include <limits>
#include <cmath>
#include <iostream>


int main ()
{
  feenableexcept(FE_DIVBYZERO|FE_INVALID);

  double x = std::numeric_limits<double>::signaling_NaN();

  if (std::isnan(x))
    std::cout << "Yes" << std::endl;
}
```

How is one supposed to test for a signalling NaN when FE_INVALID is specified?

**Comment by bangerth** [comment]
See also https://sourceware.org/bugzilla/show_bug.cgi?id=5768 .

**Comment by tjhei** [comment]
I would simply disable FPEs for those three tests.

**Comment by bangerth** [comment]
OK, will do.

**Comment by bangerth** [comment]
OK, this should now be ready. The last failure is fixed in #1218.

What would be helpful now is if we added code that initializes arrays with invalid values -- in particular with `std::numeric_limits<double>::signalling_NaN()`.

**Comment by tjhei** [comment]
A first step would be to steal the Tensor stuff from ASPECT: https://github.com/geodynamics/aspect/blob/master/include/aspect/utilities.h#L122

**Comment by bangerth** [comment]
Yes. We should take that over and put it into namespace `numbers`. The question is more: we need to identify the places where we want to initialize stuff with it.

But all of this is for another PR.

**Comment by bangerth** [comment]
Ping?

**Comment by bangerth** [comment]
Ping?

File: tests/tests.h [hunk 1]
```diff
@@ -428,10 +431,26 @@ struct SwitchOffStacktrace
   SwitchOffStacktrace ()
   {
     deal_II_exceptions::suppress_stacktrace_in_exceptions ();
-  }
+    }
```
Comment (tjhei) [inline tests/tests.h]: this indentation looks weird.

File: cmake/checks/check_02_system_features.cmake [hunk 2]
```diff
@@ -67,6 +68,41 @@ IF(NOT m_LIBRARY MATCHES "-NOTFOUND")
...
+    description << lower_bound;
+
+    return 0;
+  }
+  " 
+  _HAVE_FP_EXCEPTIONS)
+
+IF(_HAVE_FP_EXCEPTIONS)
+  MESSAGE(STATUS "Checking for Floating Point Exception macros -- Success")
+  SET(DEAL_II_HAVE_FP_EXCEPTIONS 1)
```
Comment (tjhei) [inline cmake/checks/check_02_system_features.cmake]: In ASPECT I am defining a cached variable so the user can disable it if needed. Maybe that makes sense here too? See https://github.com/geodynamics/aspect/blob/master/CMakeLists.txt#L179

**Comment by tjhei** [comment]
Looks good but see my two comments above.

File: cmake/checks/check_02_system_features.cmake (hunk 2 above)
Comment (bangerth) [inline cmake/checks/check_02_system_features.cmake]: Like so with the cached variable? (github will erase this comment in a second -- go to `check_02_system_features.cmake` around line 100.)

**Comment by tjhei** [comment]
Thanks.

**Pull request merged**